File: problem-solving/_lc-3-medium-longest-substring-without-repeating-characters/3-linear.py
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters

Complexity
----------
Time: O(n)
Space: O(n)
"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mx = 0
        mp = collections.Counter()
        left = right = 0
        while right < len(s):  # Can also be done using for loop
            mp[s[right]] += 1
            while mp[s[right]] > 1:
                mp[s[left]] -= 1
                left += 1
            mx = max(mx, right - left + 1)
            right += 1
        return mx

File: problem-solving/_lc-3-medium-longest-substring-without-repeating-characters/3-linear.go
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters

Complexity
----------
Time: O(n)
Space: O(n)
"""

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func lengthOfLongestSubstring(s string) int {
    seen := make(map[byte]int)
    ans, left, right := 0, 0, 0
    for right < len(s) {
        seen[s[right]] += 1
        for seen[s[right]] > 1 {
            seen[s[left]] -= 1
            left += 1
        }
        ans = max(ans, right - left + 1)
        right += 1
    }
    return ans
}
